src/compile/legend/parse.ts

```diff
@@ -23,8 +26,10 @@ export function parseLegend(model: Model) {
 }
 
 function parseUnitLegend(model: UnitModel): LegendComponentIndex {
-  return [COLOR, SIZE, SHAPE, OPACITY].reduce(function(legendComponent, channel) {
-    if (model.legend(channel)) {
+  const {encoding} = model;
+  return [COLOR, SIZE, SHAPE, OPACITY].reduce(function (legendComponent, channel) {
+    const def = encoding[channel];
+    if (model.legend(channel) && !(isFieldDef(def) && (channel === SHAPE && def.type === GEOJSON))) {
```

Comment:
Since you already check the same thing in scale, maybe

`model.getScaleComponent[channel] && model.legend(channel)` is sufficient?